apps/prairielearn/src/pages/instructorAssessmentSettings/instructorAssessmentSettings.js

```diff
@@ -17,147 +15,116 @@ import {
+        res.locals.assessment.tid,
+        'infoAssessment.json',
+      ),
+    );
+    res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+  }),
+);
+
+router.post(
+  '/',
+  asyncHandler(async (req, res) => {
+    if (req.body.__action === 'copy_assessment') {
+      const editor = new AssessmentCopyEditor({
+        locals: res.locals,
       });
-    });
-  } else if (req.body.__action === 'change_id') {
-    debug(`Change tid from ${res.locals.assessment.tid} to ${req.body.id}`);
-    if (!req.body.id) return next(new Error(`Invalid TID (was falsy): ${req.body.id}`));
-    if (!/^[-A-Za-z0-9_/]+$/.test(req.body.id)) {
-      return next(
-        new Error(
-          `Invalid TID (was not only letters, numbers, dashes, slashes, and underscores, with no spaces): ${req.body.id}`,
-        ),
+      const serverJob = await editor.prepareServerJob();
+      try {
+        await editor.executeWithServerJob(serverJob);
+      } catch (err) {
+        return res.redirect(res.locals.urlPrefix + '/edit_error/' + serverJob.jobSequenceId);
+      }
+
+      const assessmentId = await sqldb.queryRow(
+        sql.select_assessment_id_from_uuid,
+        { uuid: editor.uuid, course_instance_id: res.locals.course_instance.id },
+        IdSchema,
       );
-    }
-    let tid_new;
-    try {
-      tid_new = path.normalize(req.body.id);
-    } catch (err) {
-      return next(new Error(`Invalid TID (could not be normalized): ${req.body.id}`));
-    }
-    if (res.locals.assessment.tid === tid_new) {
-      debug('The new tid is the same as the old tid - do nothing');
-      res.redirect(req.originalUrl);
-    } else {
-      const editor = new AssessmentRenameEditor({
+
+      flash(
+        'success',
+        'Assessment copied successfully. You are now viewing your copy of the assessment.',
+      );
+      res.redirect(res.locals.urlPrefix + '/assessment/' + assessmentId + '/settings');
+    } else if (req.body.__action === 'delete_assessment') {
+      const editor = new AssessmentDeleteEditor({
         locals: res.locals,
-        tid_new,
       });
-
-      editor.canEdit((err) => {
-        if (ERR(err, next)) return;
-        editor.doEdit((err, job_sequence_id) => {
-          if (ERR(err, (e) => logger.error('Error in doEdit()', e))) {
-            res.redirect(res.locals.urlPrefix + '/edit_error/' + job_sequence_id);
-          } else {
-            res.redirect(req.originalUrl);
-          }
+      const serverJob = await editor.prepareServerJob();
+      try {
+        await editor.executeWithServerJob(serverJob);
+        res.redirect(res.locals.urlPrefix + '/instance_admin/assessments');
+      } catch (err) {
+        res.redirect(res.locals.urlPrefix + '/edit_error/' + serverJob.jobSequenceId);
+      }
+    } else if (req.body.__action === 'change_id') {
+      if (!req.body.id) throw new Error(`Invalid TID (was falsy): ${req.body.id}`);
+      if (!/^[-A-Za-z0-9_/]+$/.test(req.body.id)) {
+        throw new Error(
```

Comment:
Let's make these 400 errors while we're here.

Reply:
Updated to 400 errors in [c84c158](https://github.com/PrairieLearn/PrairieLearn/pull/9644/commits/c84c1586f7705e82bd635d4c553ca9ee0165f531).